# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1213208466 14400
#      Wed Jun 11 14:21:06 2008 -0400
# Node ID f74669a09debfac7ff861b211e526ffaf26306dd
# Parent  7609a985a052aa0eddd09b2f5193f85c4def08f1
rat.m: handle arrays and all-integer inputs

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,14 @@
+2008-06-11  John W. Eaton  <jwe@octave.org>
+
+	* general/rat.m: Properly initialize steps when all elements of
+	input array are integers.  Append spaces as necessary when
+	building  character array.
+
 2008-06-05  Jaroslav Hajek <highegg@gmail.com>
 
 	* plot/__go_draw_axes__.m: Use fprintf for formatted output.
 
 2008-06-04  Michael Goffioul  <michael.goffioul@gmail.com>
 
 	* plot/__go_draw_axes__.m: Support hggroup objects.
 	* plot/hggroup.m: New file.
diff --git a/scripts/general/rat.m b/scripts/general/rat.m
--- a/scripts/general/rat.m
+++ b/scripts/general/rat.m
@@ -59,18 +59,23 @@ function [n,d] = rat(x,tol)
 
   nd = ndims(y);
   nsz = numel (y);
   steps = zeros([nsz, 0]);
 
   ## grab new factors until all continued fractions converge
   while (1)
     ## determine which fractions have not yet converged
-    idx = find (abs(y-n./d) >= tol);
-    if (isempty(idx)) break; endif
+    idx = find(abs (y-n./d) >= tol);
+    if (isempty(idx))
+      if (isempty (steps))
+	steps = NaN .* ones (nsz, 1);
+      endif
+      break;
+    endif
 
     ## grab the next step in the continued fraction
     flip = 1./frac(idx);
     step = round(flip); # next element in the continued fraction
 
     if (nargout < 2)
       tsteps = NaN .* ones (nsz, 1);
       tsteps (idx) = step;
@@ -123,13 +128,20 @@ function [n,d] = rat(x,tol)
 	    s = [s(1:end-1), " + 1/", int2str(step)];
 	  endif
 	  break;
 	else
 	  s = [s(1:end-1), " + 1/(", int2str(step), ")"];
         endif
       endwhile
       s = [s, repmat(")", 1, j-2)];
+      n_nc = columns (n);
+      s_nc = columns (s);
+      if (n_nc > s_nc)
+	s(:,s_nc+1:n_nc) = " "
+      elseif (s_nc > n_nc)
+	n(:,n_nc+1:s_nc) = " ";
+      endif
       n = cat (1, n, s);
     endfor
   endif
 
 endfunction
